build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/SplitPackagesAuditTask.java

```diff
@@ -0,0 +1,324 @@
+                deps.forEach(f -> msg.add("    " + formatDependency(f)));
+                msg.add("  Classes:");
+                entry.getValue().forEach(c -> msg.add("    '" + c + "',"));
+                LOGGER.error(String.join(System.lineSeparator(), msg));
+            }
+            if (splitPackages.isEmpty() == false) {
+                throw new GradleException("Split packages found! See errors above for details.\n" +
+                    "DO NOT ADD THESE SPLIT PACKAGES TO THE IGNORE LIST! Choose a new package name for the classes added.");
+            }
+
+            try {
+                Files.write(parameters.getMarkerFile().getAsFile().get().toPath(), new byte[] {}, StandardOpenOption.CREATE);
+            } catch (IOException e) {
+                throw new RuntimeException("Failed to create marker file", e);
+            }
+        }
+
+        private Map<String, List<File>> getDependencyPackages() {
+            Map<String, List<File>> packages = new HashMap<>();
+            for (File classpathElement : getParameters().getClasspath().getFiles()) {
+                if (classpathElement.exists() == false) {
+                    continue;
+                }
+                for (String packageName : readPackages(classpathElement)) {
+                    packages.computeIfAbsent(packageName, k -> new ArrayList<>()).add(classpathElement);
+                }
+            }
+            if (LOGGER.isInfoEnabled()) {
+                List<String> msg = new ArrayList<>();
+                msg.add("Packages from dependencies:");
+                packages.entrySet().stream().sorted().forEach(e -> msg.add("  -" + e.getKey() + " -> " + e.getValue()));
+                LOGGER.info(String.join(System.lineSeparator(), msg));
+            }
+            return packages;
+        }
+
+        private Map<String, Set<String>> findSplitPackages(Set<String> dependencyPackages) {
+            Map<String, Set<String>> splitPackages = new HashMap<>();
+            for (File srcDir : getParameters().getSrcDirs().get()) {
+                try {
+                    walkJavaFiles(srcDir.toPath(), ".java", classfile -> {
+                        String packageName = getPackageName(classfile);
+                        String className = classfile.subpath(classfile.getNameCount() - 1, classfile.getNameCount()).toString();
+                        className = className.substring(0, className.length() - ".java".length());
+                        className = className.replace('$', '.'); // for inner class names
+                        LOGGER.info("Inspecting " + classfile + System.lineSeparator()
+                            + "  package: " + packageName + System.lineSeparator()
+                            + "  class: " + className);
+                        if (dependencyPackages.contains(packageName)) {
+                            splitPackages.computeIfAbsent(packageName, k -> new HashSet<>()).add(packageName + "." + className);
```

Comment:
I think we should use a `TreeSet` instead of a `HashSet` here. It might be helpful when reporting these findings that classnames are in sorted order.